src/Libraries/Microsoft.Extensions.AI/Functions/AIFunctionFactory.cs

```diff
@@ -181,35 +221,104 @@ public static ReflectionAIFunction Build(MethodInfo method, object? target, AIFu
             return new(functionDescriptor, target, options);
         }
 
+        public static ReflectionAIFunction Build(
+            MethodInfo method,
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type targetType,
+            AIFunctionFactoryOptions options)
+        {
+            _ = Throw.IfNull(method);
+
+            if (method.ContainsGenericParameters)
+            {
+                Throw.ArgumentException(nameof(method), "Open generic methods are not supported");
+            }
+
+            if (method.IsStatic)
+            {
+                Throw.ArgumentException(nameof(method), "The method must be an instance method.");
+            }
+
+            if (method.DeclaringType is { } declaringType &&
+                !declaringType.IsAssignableFrom(targetType))
+            {
+                Throw.ArgumentException(nameof(targetType), "The target type must be assignable to the method's declaring type.");
+            }
+
+            return new(ReflectionAIFunctionDescriptor.GetOrCreate(method, options), targetType, options);
+        }
+
         private ReflectionAIFunction(ReflectionAIFunctionDescriptor functionDescriptor, object? target, AIFunctionFactoryOptions options)
         {
             FunctionDescriptor = functionDescriptor;
             Target = target;
             AdditionalProperties = options.AdditionalProperties ?? EmptyReadOnlyDictionary<string, object?>.Instance;
         }
 
+        private ReflectionAIFunction(
+            ReflectionAIFunctionDescriptor functionDescriptor,
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type targetType,
+            AIFunctionFactoryOptions options)
+        {
+            FunctionDescriptor = functionDescriptor;
+            TargetType = targetType;
+            AdditionalProperties = options.AdditionalProperties ?? EmptyReadOnlyDictionary<string, object?>.Instance;
+        }
+
         public ReflectionAIFunctionDescriptor FunctionDescriptor { get; }
         public object? Target { get; }
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)]
```

Comment:
I've never seen this attribute being applied to out parameters before. What is the purpose of that?

Reply:
Wouldn't the constructor parameter annotation suffice in this case? I don't think the linker would be able to reason about the type of the returned property.

Reply:
It looks like it suffices. I think I'd added it before adding the ctor annotation.